### DISCUSSION TITLE
Warning trying to use more than one Sonoff ZBDongle-Plus

### DISCUSSION THREAD

**Discussion by shackrat** [Koenkk/zigbee2mqtt#16466] [Devices]
It appears that a batch of the Sonoff ZBDongle-P Zigbee sticks has arrived that do not have device serial numbers programmed.  This is probably not an issue for 99% of people using this device, however for those who are trying to use more than one at a time, getting both to be recognized in Linux can be can be very frustrating.

The recommended way to access serial devices in Linux is using the ID method (/dev/serial/bi-id/xxx) since those survive reboots.  Those ID's are formed in part from the onboard serial number that is flashed to the device.  The batch of devices I received all came with the serial number set to "0001".  This obviously presents a problem since UDEV rules do not assign two unique devices, instead there is only one device in the ./by-id/ directory, and it is a toss-up as to which ZBDongle-P it belongs to.

Here's an example of what one would see with a device with no real serial number programmed:

```
pi@haradiosrv:~/z2m_ch12 $ ls /dev/serial/by-id/ -n
total 0
lrwxrwxrwx 1 0 0 13 Feb  1 15:39 usb-Silicon_Labs_Sonoff_Zigbee_3.0_USB_Dongle_Plus_0001-if00-port0 -> ../../ttyUSB1
```
Note, I actually have 2 adapters attached, but since they both have serial "0001" only one appears.

Fortunately the solution is simple.  The ZBDongle-P uses a CP2102 UART chip.  SILabs offers a tool for writing settings to this chip (including device serial number) so long as the chip has not been locked by the manufacturer.  Fortunately, it appears Dongle-P's with no serial number are all unlocked (at least mine were).

To program a new serial number, you need to [download](https://community.silabs.com/s/article/cp210x-legacy-programming-utilities?language=en_US) the CP210X Legacy Utilities.  No registration is required and no installation is required, just extract the file.  (The tools are Windows-only).

Steps to fix are very simple..

1.  Launch CP210XSetIDs.exe
2. Check the box next to the serial number field.  
3. Enter something unique to that device. (i.e. "N2RWE ZBDevDevice1")
4. Click Program Device

After doing that and re-inserting the device onto the Raspberry Pi, this is what I now see in the /dev/serial/by-id/ folder:

```
pi@haradiosrv:~/z2m_ch12 $ ls /dev/serial/by-id/ -n
total 0
lrwxrwxrwx 1 0 0 13 Feb  1 15:39 usb-Silicon_Labs_Sonoff_Zigbee_3.0_USB_Dongle_Plus_0001-if00-port0 -> ../../ttyUSB1
lrwxrwxrwx 1 0 0 13 Feb  1 16:26 usb-Silicon_Labs_Sonoff_Zigbee_3.0_USB_Dongle_Plus_N2RWE-ZBDevDevice1-if00-port0 -> ../../ttyUSB4
```
Note the new device ID for the one programmed, and also note that the second dongle with serial number 0001 now appears as well.  That too can be programmed following the same steps.

**Comment by jinnatar** [comment]
For reference, if one does not have a Windows install handy, this tool is also capable of setting the serial number on Linux: https://github.com/DiUS/cp210x-cfg
